engines/sci/sound/audio32.cpp

```diff
@@ -0,0 +1,920 @@
+// AsyncEventCheck; later versions of SSCI also call it when
+// getting the playback position. This function is
+// responsible for cleaning up finished channels and
+// filling active channel buffers with decompressed audio
+// matching the hardware output audio format so they can
+// just be copied into the main DAC buffer directly later.
+//
+// The second function is called by the audio hardware when
+// the DAC buffer needs to be filled, and by `play` when
+// there is only one active sample (so it can just blow away
+// whatever was already in the DAC buffer). It merges all
+// active channels into the DAC buffer and then updates the
+// offset into the DAC buffer.
+//
+// Finally, a third function is called by the second
+// function, and it actually puts data into the DAC buffer,
+// performing volume, distortion, and balance adjustments.
+//
+// Since we only have one callback from the audio thread,
+// and should be able to do all audio processing in
+// real time, and we have streams, and we do not need to
+// completely fill the audio buffer, the functionality of
+// all these original functions is combined here and
+// simplified.
+int Audio32::readBuffer(Audio::st_sample_t *const buffer, const int numSamples) {
+	Common::StackLock lock(_mutex);
+
+	freeUnusedChannels();
+
+	// The system mixer should not try to get data when
+	// Audio32 is paused
+	assert(_pausedAtTick == 0);
+
+	// The caller of `readBuffer` is a rate converter,
+	// which reuses (without clearing) an intermediate
+	// buffer, so we need to zero the intermediate buffer
+	// to prevent mixing into audio data from the last
+	// callback.
+	memset(buffer, 0, numSamples * sizeof(Audio::st_sample_t));
+
+	// This emulates the attenuated mixing mode of SSCI
+	// engine, which halves the content of the target
+	// buffer when each new channel is mixed in.
+	// Instead of manipulating the content of the target
+	// buffer when mixing (which would either require
+	// modification of RateConverter or an expensive second
+	// pass against the entire target buffer), we just
+	// scale the volume for each channel in advance, with
+	// the earliest channel having the highest amount of
+	// attenuation (lowest volume) and the final channel
+	// having no attenuation.
+	uint8 attenuationAmount = _numActiveChannels;
+
+	int maxSamplesWritten = 0;
+
+	for (int16 channelIndex = 0; channelIndex < _numActiveChannels; ++channelIndex) {
+		--attenuationAmount;
+
+		AudioChannel &channel = getChannel(channelIndex);
+
+		if (channel.pausedAtTick || (channel.robot && _robotAudioPaused)) {
+			continue;
+		}
+
+		// Channel finished fading and had the
+		// stopChannelOnFade flag set, so no longer exists
+		if (channel.fadeStepsRemaining && processFade(channelIndex)) {
+			--channelIndex;
+			continue;
+		}
+
+		if (channel.robot) {
+			// TODO: Robot audio into output buffer
+			continue;
+		}
+
+		if (channel.vmd) {
+			// TODO: VMD audio into output buffer
```

Comment:
VMD videos are actually not played within the engine - unlike robot videos, they're full screen videos, and can be placed in specific screen areas. Thus, in ScummVM, VMD videos run within the video decoder interface, so their audio handling occurs outside the engine. Thus, this VMD audio handling code is probably not needed with our implementation